runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/BeamFnMapTaskExecutor.java

```diff
@@ -351,7 +398,106 @@ void updateProgress() {
+    // when more counter types are added.
+    // todomigryz: define counter transformer factory
+    // that can provide respective counter transformer for different type of counters.
+    // (ie RowCountCounterTranformer, MSecCounterTransformer, UserCounterTransformer, etc)
+    private static class MonitoringInfoToCounterUpdateTransformer {
+
+      private final Map<String, DataflowStepContext> transformIdMapping;
+
+      public MonitoringInfoToCounterUpdateTransformer(
+          final Map<String, DataflowStepContext> transformIdMapping) {
+        this.transformIdMapping = transformIdMapping;
+      }
+
+      // todo: search code for "beam:metrics"... and replace them with relevant enums from
+      // proto after rebasing above https://github.com/apache/beam/pull/6799 that
+      // introduces relevant proto entries.
+      final String BEAM_METRICS_USER_PREFIX = "beam:metric:user";
+
+      private CounterUpdate monitoringInfoToCounterUpdate(MonitoringInfo monitoringInfo) {
+        long value = monitoringInfo.getMetric().getCounterData().getInt64Value();
+        String urn = monitoringInfo.getUrn();
+
+        String type = monitoringInfo.getType();
+
+        // todo: run MonitoringInfo through validation process.
+        // refer to https://github.com/apache/beam/pull/6799
+
+        if (urn.startsWith(BEAM_METRICS_USER_PREFIX)) {
+          if (!type.equals("beam:metrics:sum_int_64")) {
+            return null;
+          }
+
+          final String ptransform = monitoringInfo.getLabelsMap().get("PTRANSFORM");
+          if (ptransform == null) {
+            return null;
+          }
+
+          DataflowStepContext stepContext = transformIdMapping.get(ptransform);
+
+          CounterStructuredNameAndMetadata name = new CounterStructuredNameAndMetadata();
+
+          String nameWithNamespace =
+              monitoringInfo
+                  .getUrn()
+                  .substring(BEAM_METRICS_USER_PREFIX.length())
+                  .replace("^:", "");
+
+          final int lastColonIndex = nameWithNamespace.lastIndexOf(':');
+          String counterName = nameWithNamespace.substring(lastColonIndex + 1);
+          String counterNamespace = nameWithNamespace.substring(0, lastColonIndex);
+
+          name.setName(
+                  new CounterStructuredName()
+                      .setOrigin(Origin.USER.toString())
+                      // Workaround for bug in python sdk that missed colon after ...metric:user.
+                      .setName(counterName)
+                      .setOriginalStepName(
+                          stepContext == null ? null : stepContext.getNameContext().originalName())
+                      .setExecutionStepName(
+                          stepContext == null ? null : stepContext.getNameContext().systemName())
+                      .setOriginNamespace(counterNamespace))
+              .setMetadata(new CounterMetadata().setKind("SUM"));
+
+          return new CounterUpdate()
```

Comment:
This is okay for now, since you are just unpacking the user counter metric. but I want to see this share more code for different metrics later on.

Any label extraction can be shared for example.

Reply:
That is a valid point. But I'd prefer to implement this in next PR when I add other type of counters.